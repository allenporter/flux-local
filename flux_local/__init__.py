"""
.. include:: ../README.md
"""

__all__ = [
    "git_repo",
    "manifest",
    "kustomize",
    "helm",
    "exceptions",
]
