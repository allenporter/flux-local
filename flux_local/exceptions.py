"""Exceptions related to flux-local."""

__all__ = [
    "FluxException",
    "InputException",
    "CommandException",
]


class FluxException(Exception):
    """Generic base exception used for this library."""


class InputException(FluxException):
    """Raised when the input files or values are not formatted as expected."""


class CommandException(FluxException):
    """Raised when there is a failure running a subcommand."""


class KustomizeException(CommandException):
    """Raised when there is a failure running a kustomize command."""


class KustomizePathException(CommandException):
    """Raised a Kustomization points to a path that does not exist."""


class HelmException(CommandException):
    """Raised when there is a failure running a helm command."""


class KyvernoException(CommandException):
    """Raised when there is an error running kyverno policy command."""
