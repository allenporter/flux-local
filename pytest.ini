[pytest]
log_level = DEBUG
asyncio_mode = auto
