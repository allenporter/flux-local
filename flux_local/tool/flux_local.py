"""Command line tool for building, diffing, validating flux local repositories."""

import argparse
import asyncio
import logging
import pathlib
import sys

import yaml

from . import build, diff, get, manifest, test
from flux_local import command

_LOGGER = logging.getLogger(__name__)


def main() -> None:
    """Flux-local command line tool main entry point."""
    # https://github.com/yaml/pyyaml/issues/89
    yaml.Loader.yaml_implicit_resolvers.pop("=")

    parser = argparse.ArgumentParser(
        description="Command line utility for inspecting a local flux repository.",
    )
    parser.add_argument(
        "--log-level", choices=["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"]
    )

    subparsers = parser.add_subparsers(dest="command", help="Command", required=True)

    build_args = subparsers.add_parser(
        "build",
        help="Build local flux Kustomization target from a local directory",
    )
    build_args.add_argument(
        "path", type=pathlib.Path, help="Path to the kustomization or charts"
    )
    build_args.add_argument(
        "--enable-helm",
        type=bool,
        action=argparse.BooleanOptionalAction,
        help="Enable use of HelmRelease inflation",
    )
    # pylint: disable=duplicate-code
    build_args.add_argument(
        "--skip-crds",
        type=bool,
        default=False,
        action=argparse.BooleanOptionalAction,
        help="Allows disabling of outputting CRDs to reduce output size",
    )
    build_args.set_defaults(cls=build.BuildAction)

    manifest_args = subparsers.add_parser(
        "manifest", help="Build a yaml manifest file representing the cluster"
    )
    manifest_args.add_argument(
        "path", type=pathlib.Path, help="Path to the kustomization or charts"
    )
    manifest_args.set_defaults(cls=manifest.ManifestAction)

    test_args = subparsers.add_parser("test", help="Build and validate the cluster")
    test_args.add_argument(
        "path", type=pathlib.Path, help="Path to the kustomization or charts"
    )
    test_args.set_defaults(cls=test.TestAction)

    get.GetAction.register(subparsers)
    diff.DiffAction.register(subparsers)

    args = parser.parse_args()

    if args.log_level:
        logging.basicConfig(level=args.log_level)

    action = args.cls()
    try:
        asyncio.run(action.run(**vars(args)))
    except command.CommandException as err:
        print("Command failed: ", err)
        sys.exit(1)


if __name__ == "__main__":
    main()
