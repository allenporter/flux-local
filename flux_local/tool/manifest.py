"""Flux-local manifest action."""

import pathlib

from flux_local import git_repo


class ManifestAction:
    """Flux-local manifest action."""

    async def run(  # type: ignore[no-untyped-def]
        self,
        path: pathlib.Path,
        **kwargs,  # pylint: disable=unused-argument
    ) -> None:
        """Async Action implementation."""
        manifest = await git_repo.build_manifest(path)
        print(manifest.yaml())
