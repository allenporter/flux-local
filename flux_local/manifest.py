"""Representation of the contents of a cluster.

A manifest may be built directly from the local context of a cluster, or may be
serialized and stored and checked into the cluster for use in other applications
e.g. such as writing management plan for resources.
"""

from pathlib import Path
from typing import Any, Optional, cast

import aiofiles
import yaml

try:
    from pydantic.v1 import BaseModel, Field
except ImportError:
    from pydantic import BaseModel, Field  # type: ignore

from .exceptions import InputException

__all__ = [
    "read_manifest",
    "write_manifest",
    "Manifest",
    "Cluster",
    "Kustomization",
    "HelmRepository",
    "HelmRelease",
    "HelmChart",
    "ClusterPolicy",
]

# Match a prefix of apiVersion to ensure we have the right type of object.
# We don't check specific versions for forward compatibility on upgrade.
FLUXTOMIZE_DOMAIN = "kustomize.toolkit.fluxcd.io"
KUSTOMIZE_DOMAIN = "kustomize.config.k8s.io"
HELM_REPO_DOMAIN = "source.toolkit.fluxcd.io"
HELM_RELEASE_DOMAIN = "helm.toolkit.fluxcd.io"
CLUSTER_POLICY_DOMAIN = "kyverno.io"
CRD_KIND = "CustomResourceDefinition"
SECRET_KIND = "Secret"

REPO_TYPE_DEFAULT = "default"
REPO_TYPE_OCI = "oci"


def _check_version(doc: dict[str, Any], version: str) -> None:
    """Assert that the resource has the specified version."""
    if not (api_version := doc.get("apiVersion")):
        raise InputException(f"Invalid object missing apiVersion: {doc}")
    if not api_version.startswith(version):
        raise InputException(f"Invalid object expected '{version}': {doc}")


class BaseManifest(BaseModel):
    """Base class for all manifest objects."""

    _COMPACT_EXCLUDE_FIELDS: dict[str, Any] = {}

    def compact_dict(self, exclude: dict[str, Any] | None = None) -> dict[str, Any]:
        """Return a compact dictionary representation of the object.

        This is similar to `dict()` but with a specific implementation for serializing
        with variable fields removed.
        """
        if exclude is None:
            exclude = self._COMPACT_EXCLUDE_FIELDS
        return self.dict(exclude=exclude)  # type: ignore[arg-type]

    @classmethod
    def parse_yaml(cls, content: str) -> "BaseManifest":
        """Parse a serialized manifest."""
        doc = next(yaml.load_all(content, Loader=yaml.Loader), None)
        return cls.parse_obj(doc)

    def yaml(self, exclude: dict[str, Any] | None = None) -> str:
        """Return a YAML string representation of compact_dict."""
        data = self.compact_dict(exclude)
        return yaml.dump(data, sort_keys=False, explicit_start=True)


class HelmChart(BaseManifest):
    """A representation of an instantiation of a chart for a HelmRelease."""

    name: str
    """The name of the chart within the HelmRepository."""

    version: Optional[str] = None
    """The version of the chart."""

    repo_name: str
    """The name of the HelmRepository."""

    repo_namespace: str
    """The namespace of the HelmRepository."""

    @classmethod
    def parse_doc(cls, doc: dict[str, Any], default_namespace: str) -> "HelmChart":
        """Parse a HelmChart from a HelmRelease resource object."""
        _check_version(doc, HELM_RELEASE_DOMAIN)
        if not (spec := doc.get("spec")):
            raise InputException(f"Invalid {cls} missing spec: {doc}")
        if not (chart := spec.get("chart")):
            raise InputException(f"Invalid {cls} missing spec.chart: {doc}")
        if not (chart_spec := chart.get("spec")):
            raise InputException(f"Invalid {cls} missing spec.chart.spec: {doc}")
        if not (chart := chart_spec.get("chart")):
            raise InputException(f"Invalid {cls} missing spec.chart.spec.chart: {doc}")
        version = chart_spec.get("version")
        if not (source_ref := chart_spec.get("sourceRef")):
            raise InputException(
                f"Invalid {cls} missing spec.chart.spec.sourceRef: {doc}"
            )
        if "name" not in source_ref:
            raise InputException(f"Invalid {cls} missing sourceRef fields: {doc}")
        return cls(
            name=chart,
            version=version,
            repo_name=source_ref["name"],
            repo_namespace=source_ref.get("namespace", default_namespace),
        )

    @property
    def chart_name(self) -> str:
        """Identifier for the HelmChart."""
        return f"{self.repo_namespace}-{self.repo_name}/{self.name}"

    _COMPACT_EXCLUDE_FIELDS = {"version": True}


class HelmRelease(BaseManifest):
    """A representation of a Flux HelmRelease."""

    name: str
    """The name of the HelmRelease."""

    namespace: str
    """The namespace that owns the HelmRelease."""

    chart: HelmChart
    """A mapping to a specific helm chart for this HelmRelease."""

    values: Optional[dict[str, Any]] = None
    """The values to install in the chart."""

    @classmethod
    def parse_doc(cls, doc: dict[str, Any]) -> "HelmRelease":
        """Parse a HelmRelease from a kubernetes resource object."""
        _check_version(doc, HELM_RELEASE_DOMAIN)
        if not (metadata := doc.get("metadata")):
            raise InputException(f"Invalid {cls} missing metadata: {doc}")
        if not (name := metadata.get("name")):
            raise InputException(f"Invalid {cls} missing metadata.name: {doc}")
        if not (namespace := metadata.get("namespace")):
            raise InputException(f"Invalid {cls} missing metadata.namespace: {doc}")
        chart = HelmChart.parse_doc(doc, namespace)
        return cls(
            name=name,
            namespace=namespace,
            chart=chart,
            values=doc["spec"].get("values"),
        )

    @property
    def release_name(self) -> str:
        """Identifier for the HelmRelease."""
        return f"{self.namespace}-{self.name}"

    @property
    def repo_name(self) -> str:
        """Identifier for the HelmRepository identified in the HelmChart."""
        return f"{self.chart.repo_namespace}-{self.chart.repo_name}"

    _COMPACT_EXCLUDE_FIELDS = {
        "values": True,
        "chart": HelmChart._COMPACT_EXCLUDE_FIELDS,
    }


class HelmRepository(BaseManifest):
    """A representation of a flux HelmRepository."""

    name: str
    """The name of the HelmRepository."""

    namespace: str
    """The namespace of owning the HelmRepository."""

    url: str
    """The URL to the repository of helm charts."""

    repo_type: str | None = None
    """The type of the HelmRepository."""

    @classmethod
    def parse_doc(cls, doc: dict[str, Any]) -> "HelmRepository":
        """Parse a HelmRepository from a kubernetes resource."""
        _check_version(doc, HELM_REPO_DOMAIN)
        if not (metadata := doc.get("metadata")):
            raise InputException(f"Invalid {cls} missing metadata: {doc}")
        if not (name := metadata.get("name")):
            raise InputException(f"Invalid {cls} missing metadata.name: {doc}")
        if not (namespace := metadata.get("namespace")):
            raise InputException(f"Invalid {cls} missing metadata.namespace: {doc}")
        if not (spec := doc.get("spec")):
            raise InputException(f"Invalid {cls} missing spec: {doc}")
        if not (url := spec.get("url")):
            raise InputException(f"Invalid {cls} missing spec.url: {doc}")
        return cls(
            name=name,
            namespace=namespace,
            url=url,
            repo_type=spec.get("type", REPO_TYPE_DEFAULT),
        )

    @property
    def repo_name(self) -> str:
        """Identifier for the HelmRepository."""
        return f"{self.namespace}-{self.name}"


class ClusterPolicy(BaseManifest):
    """A kyverno policy object."""

    name: str
    """The name of the kustomization."""

    namespace: str | None = None
    """The namespace of the kustomization."""

    doc: dict[str, Any] | None = None
    """The raw ClusterPolicy document."""

    @classmethod
    def parse_doc(cls, doc: dict[str, Any]) -> "ClusterPolicy":
        """Parse a cluster policy object from a kubernetes resource."""
        _check_version(doc, CLUSTER_POLICY_DOMAIN)
        if not (metadata := doc.get("metadata")):
            raise InputException(f"Invalid {cls} missing metadata: {doc}")
        if not (name := metadata.get("name")):
            raise InputException(f"Invalid {cls} missing metadata.name: {doc}")
        namespace = metadata.get("namespace")
        if not doc.get("spec"):
            raise InputException(f"Invalid {cls} missing spec: {doc}")
        return ClusterPolicy(name=name, namespace=namespace, doc=doc)

    _COMPACT_EXCLUDE_FIELDS = {
        "doc": True,
    }


class Kustomization(BaseManifest):
    """A Kustomization is a set of declared cluster artifacts.

    This represents a flux Kustomization that points to a path that
    contains typical `kustomize` Kustomizations on local disk that
    may be flat or contain overlays.
    """

    name: str
    """The name of the kustomization."""

    namespace: str | None = None
    """The namespace of the kustomization."""

    path: str
    """The local repo path to the kustomization contents."""

    helm_repos: list[HelmRepository] = Field(default_factory=list)
    """The set of HelmRepositories represented in this kustomization."""

    helm_releases: list[HelmRelease] = Field(default_factory=list)
    """The set of HelmRelease represented in this kustomization."""

    cluster_policies: list[ClusterPolicy] = Field(default_factory=list)
    """The set of ClusterPolicies represented in this kustomization."""

    source_path: str | None = None
    """Optional source path for this Kustomization, relative to the build path."""

    source_kind: str | None = None
    """The sourceRef kind that provides this Kustomization e.g. GitRepository etc."""

    source_name: str | None = None
    """The name of the sourceRef that provides this Kustomization."""

    source_namespace: str | None = None
    """The namespace of the sourceRef that provides this Kustomization."""

    @classmethod
    def parse_doc(cls, doc: dict[str, Any]) -> "Kustomization":
        """Parse a partial Kustomization from a kubernetes resource."""
        _check_version(doc, FLUXTOMIZE_DOMAIN)
        if not (metadata := doc.get("metadata")):
            raise InputException(f"Invalid {cls} missing metadata: {doc}")
        if not (name := metadata.get("name")):
            raise InputException(f"Invalid {cls} missing metadata.name: {doc}")
        if not (namespace := metadata.get("namespace")):
            raise InputException(f"Invalid {cls} missing metadata.namespace: {doc}")
        if not (spec := doc.get("spec")):
            raise InputException(f"Invalid {cls} missing spec: {doc}")
        path = spec.get("path", "")
        source_path = metadata.get("annotations", {}).get("config.kubernetes.io/path")
        source_ref = spec.get("sourceRef", {})
        return Kustomization(
            name=name,
            namespace=namespace,
            path=path,
            source_path=source_path,
            source_kind=source_ref.get("kind"),
            source_name=source_ref.get("name"),
            source_namespace=source_ref.get("namespace", namespace),
        )

    @property
    def id_name(self) -> str:
        """Identifier for the Kustomization in tests"""
        return f"{self.path}"

    @property
    def namespaced_name(self, sep: str = "/") -> str:
        """Return the namespace and name concatenated as an id."""
        return f"{self.namespace}{sep}{self.name}"

    _COMPACT_EXCLUDE_FIELDS = {
        "helm_releases": {
            "__all__": HelmRelease._COMPACT_EXCLUDE_FIELDS,
        },
        "cluster_policies": {
            "__all__": ClusterPolicy._COMPACT_EXCLUDE_FIELDS,
        },
        "source_path": True,
        "source_name": True,
        "source_namespace": True,
        "source_kind": True,
    }


class Cluster(BaseManifest):
    """A set of nodes that run containerized applications.

    Many flux git repos will only have a single flux cluster, though
    a repo may also contain multiple (e.g. dev an prod).
    """

    name: str
    """The name of the cluster."""

    namespace: str
    """The namespace of the cluster."""

    path: str
    """The local git repo path to the Kustomization objects for the cluster."""

    kustomizations: list[Kustomization] = Field(default_factory=list)
    """A list of flux Kustomizations for the cluster."""

    @classmethod
    def parse_doc(cls, doc: dict[str, Any]) -> "Cluster":
        """Parse a partial Kustomization from a kubernetes resource."""
        _check_version(doc, FLUXTOMIZE_DOMAIN)
        if not (metadata := doc.get("metadata")):
            raise InputException(f"Invalid {cls} missing metadata: {doc}")
        if not (name := metadata.get("name")):
            raise InputException(f"Invalid {cls} missing metadata.name: {doc}")
        if not (namespace := metadata.get("namespace")):
            raise InputException(f"Invalid {cls} missing metadata.namespace: {doc}")
        if not (spec := doc.get("spec")):
            raise InputException(f"Invalid {cls} missing spec: {doc}")
        if not (path := spec.get("path")):
            raise InputException(f"Invalid {cls} missing spec.path: {doc}")
        return Cluster(name=name, namespace=namespace, path=path)

    @property
    def namespaced_name(self, sep: str = "/") -> str:
        """Return the namespace and name concatenated as an id."""
        return f"{self.namespace}{sep}{self.name}"

    @property
    def id_name(self) -> str:
        """Identifier for the Cluster in tests."""
        return f"{self.path}"

    @property
    def helm_repos(self) -> list[HelmRepository]:
        """Return the list of HelmRepository objects from all Kustomizations."""
        return [
            repo
            for kustomization in self.kustomizations
            for repo in kustomization.helm_repos
        ]

    @property
    def helm_releases(self) -> list[HelmRelease]:
        """Return the list of HelmRelease objects from all Kustomizations."""
        return [
            release
            for kustomization in self.kustomizations
            for release in kustomization.helm_releases
        ]

    @property
    def cluster_policies(self) -> list[ClusterPolicy]:
        """Return the list of ClusterPolicy objects from all Kustomizations."""
        return [
            policy
            for kustomization in self.kustomizations
            for policy in kustomization.cluster_policies
        ]

    _COMPACT_EXCLUDE_FIELDS = {
        "kustomizations": {
            "__all__": Kustomization._COMPACT_EXCLUDE_FIELDS,
        }
    }


class Manifest(BaseManifest):
    """Holds information about cluster and applications contained in a repo."""

    clusters: list[Cluster]
    """A list of Clusters represented in the repo."""

    _COMPACT_EXCLUDE_FIELDS = {
        "clusters": {
            "__all__": Cluster._COMPACT_EXCLUDE_FIELDS,
        }
    }


async def read_manifest(manifest_path: Path) -> Manifest:
    """Return the contents of a serialized manifest file.

    A manifest file is typically created by `flux-local get cluster -o yaml` or
    similar command.
    """
    async with aiofiles.open(str(manifest_path)) as manifest_file:
        content = await manifest_file.read()
        if not content:
            raise ValueError("validation error for Manifest file {manifest_path}")
        return cast(Manifest, Manifest.parse_yaml(content))


async def write_manifest(manifest_path: Path, manifest: Manifest) -> None:
    """Write the specified manifest content to disk."""
    content = manifest.yaml()
    async with aiofiles.open(str(manifest_path), mode="w") as manifest_file:
        await manifest_file.write(content)


async def update_manifest(manifest_path: Path, manifest: Manifest) -> None:
    """Write the specified manifest only if changed."""
    new_content = manifest.yaml()
    async with aiofiles.open(str(manifest_path)) as manifest_file:
        content = await manifest_file.read()
    if content == new_content:
        return
    async with aiofiles.open(str(manifest_path), mode="w") as manifest_file:
        await manifest_file.write(new_content)
