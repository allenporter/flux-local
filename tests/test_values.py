"""Tests for helm library."""

import base64
from pathlib import Path

import pytest
from syrupy.assertion import SnapshotAssertion

from flux_local.exceptions import HelmException
from flux_local.values import (
    expand_value_references,
    expand_postbuild_substitute_reference,
    ks_cluster_config,
    cluster_config,
)
from flux_local.manifest import (
    HelmRelease,
    ValuesReference,
    Secret,
    ConfigMap,
    Kustomization,
    HelmChart,
    SubstituteReference,
)
from flux_local.git_repo import ResourceSelector, PathSelector, build_manifest


async def test_value_references(snapshot: SnapshotAssertion) -> None:
    """Test for expanding value references."""
    path = Path("tests/testdata/cluster8")
    selector = ResourceSelector(path=PathSelector(path=path))
    manifest = await build_manifest(selector=selector)
    assert len(manifest.clusters) == 1
    assert len(manifest.clusters[0].kustomizations) == 2
    ks = manifest.clusters[0].kustomizations[0]
    assert ks.name == "apps"
    assert len(ks.helm_releases) == 1
    hr = ks.helm_releases[0]
    assert hr.name == "podinfo"
    assert hr.values == snapshot


def test_values_references_with_values_key() -> None:
    """Test for expanding a value reference with a values key."""
    hr = HelmRelease(
        name="test",
        namespace="test",
        chart=HelmChart(
            repo_name="test-repo",
            repo_namespace="flux-system",
            name="test-chart",
            version="test-version",
        ),
        values={"test": "test"},
        values_from=[
            ValuesReference(
                kind="ConfigMap",
                name="test-values-configmap",
                values_key="some-key",
                target_path="target.path",
            ),
            ValuesReference(
                kind="ConfigMap",
                name="test-binary-data-configmap",
                values_key="some-key",
            ),
        ],
    )
    ks = Kustomization(
        name="test",
        namespace="test",
        path="example/path",
        helm_releases=[hr],
        config_maps=[
            ConfigMap(
                name="test-values-configmap",
                namespace="test",
                data={"some-key": "example_value"},
            ),
            ConfigMap(
                name="test-binary-data-configmap",
                namespace="test",
                binary_data={
                    "some-key": base64.b64encode(
                        "encoded_key: encoded_value".encode("utf-8")
                    )
                },
            ),
        ],
    )
    updated_hr = expand_value_references(hr, ks)
    assert updated_hr.values == {
        "test": "test",
        "target": {
            "path": "example_value",
        },
        "encoded_key": "encoded_value",
    }


def test_values_references_with_missing_values_key() -> None:
    """Test for expanding a value reference with a values key that is missing."""
    hr = HelmRelease(
        name="test",
        namespace="test",
        chart=HelmChart(
            repo_name="test-repo",
            repo_namespace="flux-system",
            name="test-chart",
            version="test-version",
        ),
        values={"test": "test"},
        values_from=[
            ValuesReference(
                kind="ConfigMap",
                name="test-values-configmap",
                values_key="some-key-does-not-exist",
                target_path="target.path",
            )
        ],
    )
    ks = Kustomization(
        name="test",
        namespace="test",
        path="example/path",
        helm_releases=[hr],
        config_maps=[
            ConfigMap(
                name="test-values-configmap",
                namespace="test",
                data={"some-key": "example_value"},
            )
        ],
    )
    updated_hr = expand_value_references(hr, ks)
    assert updated_hr.values == {
        "test": "test",
    }



def test_values_references_with_missing_secret() -> None:
    """Test for expanding a value reference with a missing secret."""
    hr = HelmRelease(
        name="test",
        namespace="test",
        chart=HelmChart(
            repo_name="test-repo",
            repo_namespace="flux-system",
            name="test-chart",
            version="test-version",
        ),
        values={"test": "test"},
        values_from=[
            ValuesReference(
                kind="Secret",
                name="test-values-secret",
                values_key="some-key-does-not-exist",
                target_path="target.path",
            )
        ],
    )
    ks = Kustomization(
        name="test",
        namespace="test",
        path="example/path",
        helm_releases=[hr],
    )
    updated_hr = expand_value_references(hr, ks)
    assert updated_hr.values == {
        "test": "test",
        "target": {
            "path": "!!PLACEHOLDER!!",
        },
    }


def test_values_references_with_missing_secret_values_key() -> None:
    """Test for expanding a value reference with a secret key that is missing."""
    hr = HelmRelease(
        name="test",
        namespace="test",
        chart=HelmChart(
            repo_name="test-repo",
            repo_namespace="flux-system",
            name="test-chart",
            version="test-version",
        ),
        values={"test": "test"},
        values_from=[
            ValuesReference(
                kind="Secret",
                name="test-values-secret",
                values_key="some-key-does-not-exist",
                target_path="target.path",
            )
        ],
    )
    ks = Kustomization(
        name="test",
        namespace="test",
        path="example/path",
        helm_releases=[hr],
        secrets=[
            Secret(
                name="test-values-secret",
                namespace="test",
                string_data={"some-key": "example_value"},
            )
        ],
    )
    updated_hr = expand_value_references(hr, ks)
    assert updated_hr.values == {
        "test": "test",
    }


def test_values_references_invalid_yaml() -> None:
    """Test for expanding a value reference with a values key."""
    hr = HelmRelease(
        name="test",
        namespace="test",
        chart=HelmChart(
            repo_name="test-repo",
            repo_namespace="flux-system",
            name="test-chart",
            version="test-version",
        ),
        values={"test": "test"},
        values_from=[
            ValuesReference(
                kind="ConfigMap",
                name="test-values-configmap",
            )
        ],
    )
    ks = Kustomization(
        name="test",
        namespace="test",
        path="example/path",
        helm_releases=[hr],
        config_maps=[
            ConfigMap(
                name="test-values-configmap",
                namespace="test",
                data={"values.yaml": "not-yaml"},
            )
        ],
    )
    with pytest.raises(HelmException, match=r"valid yaml"):
        expand_value_references(hr, ks)


def test_values_references_invalid_binary_data() -> None:
    """Test for expanding a value reference with an invalid binary data key."""
    hr = HelmRelease(
        name="test",
        namespace="test",
        chart=HelmChart(
            repo_name="test-repo",
            repo_namespace="flux-system",
            name="test-chart",
            version="test-version",
        ),
        values={"test": "test"},
        values_from=[
            ValuesReference(
                kind="ConfigMap",
                name="test-values-configmap",
            )
        ],
    )
    ks = Kustomization(
        name="test",
        namespace="test",
        path="example/path",
        helm_releases=[hr],
        config_maps=[
            ConfigMap(
                name="test-values-configmap",
                namespace="test",
                binary_data={"values.yaml": "this is not base64 data"},
            )
        ],
    )
    with pytest.raises(HelmException, match=r"Unable to decode binary data"):
        expand_value_references(hr, ks)


def test_values_reference_invalid_target_path() -> None:
    """Test for expanding a value reference with a values key."""
    hr = HelmRelease(
        name="test",
        namespace="test",
        chart=HelmChart(
            repo_name="test-repo",
            repo_namespace="flux-system",
            name="test-chart",
            version="test-version",
        ),
        values={
            "test": "test",
            "target": ["a", "b", "c"],
        },
        values_from=[
            ValuesReference(
                kind="ConfigMap",
                name="test-values-configmap",
                values_key="some-key",
                # Target above is a list
                target_path="target.path",
            )
        ],
    )
    ks = Kustomization(
        name="test",
        namespace="test",
        path="example/path",
        helm_releases=[hr],
        config_maps=[
            ConfigMap(
                name="test-values-configmap",
                namespace="test",
                data={"some-key": "example_value"},
            )
        ],
    )
    with pytest.raises(HelmException, match=r"values to be a dict"):
        expand_value_references(hr, ks)


def test_values_reference_invalid_configmap_and_secret() -> None:
    """Test a values reference to a config map and secret that do not exist."""
    hr = HelmRelease(
        name="test",
        namespace="test",
        chart=HelmChart(
            repo_name="test-repo",
            repo_namespace="flux-system",
            name="test-chart",
            version="test-version",
        ),
        values={"test": "test"},
        values_from=[
            ValuesReference(
                kind="ConfigMap",
                name="test-values-configmap",
                optional=False,  # We just log
            ),
            ValuesReference(
                kind="Secret",
                name="test-values-secret",
                optional=False,  # We just log
            ),
            ValuesReference(
                kind="UnknownKind",
                name="test-values-secret",
            ),
        ],
    )
    ks = Kustomization(
        name="test",
        namespace="test",
        path="example/path",
        helm_releases=[hr],
        config_maps=[],
    )
    updated_hr = expand_value_references(hr, ks)
    # No changes to the values
    assert updated_hr.values == {"test": "test"}


def test_values_references_secret() -> None:
    """Test for expanding a value reference for a secret."""
    hr = HelmRelease(
        name="test",
        namespace="test",
        chart=HelmChart(
            repo_name="test-repo",
            repo_namespace="flux-system",
            name="test-chart",
            version="test-version",
        ),
        values={"test": "test"},
        values_from=[
            ValuesReference(
                kind="Secret",
                name="test-values-secret",
                values_key="some-key1",
                target_path="target.path1",
            ),
            ValuesReference(
                kind="Secret",
                name="test-string-values-secret",
                values_key="some-key2",
                target_path="target.path2",
            ),
        ],
    )
    ks = Kustomization(
        name="test",
        namespace="test",
        path="example/path",
        helm_releases=[hr],
        secrets=[
            Secret.parse_doc(
                {
                    "apiVersion": "v1",
                    "kind": "Secret",
                    "metadata": {
                        "name": "test-values-secret",
                        "namespace": "test",
                    },
                    "data": {
                        "some-key1": base64.b64encode("example-value".encode("utf-8")),
                    },
                }
            ),
            Secret.parse_doc(
                {
                    "apiVersion": "v1",
                    "kind": "Secret",
                    "metadata": {
                        "name": "test-string-values-secret",
                        "namespace": "test",
                    },
                    "stringData": {
                        "some-key2": "example-string-value",
                    },
                }
            ),
        ],
    )
    updated_hr = expand_value_references(hr, ks)
    assert updated_hr.values == {
        "test": "test",
        "target": {
            "path1": "!!PLACEHOLDER!!",
            "path2": "!!PLACEHOLDER!!",
        },
    }


def test_postbuild_substitute_from() -> None:
    """Test for expanding from a config map."""
    app_ks = Kustomization(
        name="app",
        namespace="test",
        path="example/app",
        postbuild_substitute={"test": "test"},
        postbuild_substitute_from=[
            SubstituteReference(
                kind="ConfigMap",
                name="test-values-configmap",
            ),
            SubstituteReference(
                kind="ConfigMap",
                name="test-binary-data-configmap",
            ),
        ],
    )
    config_maps = [
        ConfigMap(
            name="test-values-configmap",
            namespace="test",
            data={"some_key": "example_value"},
        ),
        ConfigMap(
            name="test-binary-data-configmap",
            namespace="test",
            binary_data={
                "encoded_key": base64.b64encode("encoded_value".encode("utf-8"))
            },
        ),
    ]
    app_ks = expand_postbuild_substitute_reference(
        app_ks, cluster_config([], config_maps)
    )
    assert app_ks.postbuild_substitute == {
        "test": "test",
        "some_key": "example_value",
        "encoded_key": "encoded_value",
    }


def test_postbuild_substitute_from_secret() -> None:
    """Test for expanding from a secret."""
    app_ks = Kustomization(
        name="app",
        namespace="test",
        path="example/app",
        postbuild_substitute={"test": "test"},
        postbuild_substitute_from=[
            SubstituteReference(
                kind="Secret",
                name="test-values-secret",
            ),
        ],
    )
    secrets = [
        Secret(
            name="test-values-secret",
            namespace="test",
            string_data={"some_key": "example_value"},
        ),
    ]
    app_ks = expand_postbuild_substitute_reference(app_ks, cluster_config(secrets, []))
    assert app_ks.postbuild_substitute == {
        "test": "test",
        "some_key": "example_value",
    }


def test_postbuild_substitute_from_invalid_references() -> None:
    """Test for invalid SubstituteReferences."""
    app_ks = Kustomization(
        name="app",
        namespace="test",
        path="example/app",
        postbuild_substitute={"test": "test"},
        postbuild_substitute_from=[
            SubstituteReference(
                kind="Secret",
                name="test-values-secret",
            ),
            SubstituteReference(
                kind="ConfigMap",
                name="test-values-config-map",
            ),
            SubstituteReference(
                kind="Invalid",
                name="test-values-invalid",
            ),
        ],
    )
    app_ks = expand_postbuild_substitute_reference(app_ks, cluster_config([], []))
    assert app_ks.postbuild_substitute == {
        "test": "test",
    }


def test_cluster_config() -> None:
    """Test for ClusterConfig objects."""

    ks1 = Kustomization(
        name="app",
        namespace="test",
        path="example/app",
        secrets=[
            Secret(
                name="test-values-secret",
                namespace="test",
                string_data={"some_key": "example_value"},
            ),
        ],
    )
    ks2 = Kustomization(
        name="app",
        namespace="test",
        path="example/app",
        secrets=[
            Secret(
                name="other-values-secret",
                namespace="test",
                string_data={"some_key": "example_value"},
            ),
        ],
    )
    config = ks_cluster_config([ks1, ks2])
    # Ensure can be called repeatedly
    assert len(list(config.secrets)) == 2
    assert [ s.name for s in config.secrets ] == ['test-values-secret','other-values-secret']
