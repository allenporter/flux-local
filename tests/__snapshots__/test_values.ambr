# serializer version: 1
# name: test_value_references
  dict({
    'ingress': dict({
      'className': 'nginx',
      'enabled': True,
      'hosts': list([
        dict({
          'host': 'podinfo.production',
          'paths': list([
            dict({
              'path': '/',
              'pathType': 'ImplementationSpecific',
            }),
          ]),
        }),
      ]),
    }),
    'redis': dict({
      'enabled': True,
      'repository': 'public.ecr.aws/docker/library/redis',
      'tag': '7.0.6',
    }),
    'tls': dict({
      'crt': '..PLACEHOLDER..',
    }),
  })
# ---
# name: test_value_references.1
  dict({
    'apiServerProxyConfig': dict({
      'mode': 'true',
    }),
    'oauth': dict({
      'clientId': '..PLACEHOLDER..',
      'clientSecret': '..PLACEHOLDER..',
    }),
    'operatorConfig': dict({
      'defaultTags': list([
        'tag:k8s',
      ]),
      'hostname': 'tailscale-operator',
    }),
  })
# ---
